src/main/java/gyro/google/kms/CryptoKeyResource.java

```diff
@@ -0,0 +1,382 @@
+        return cryptoKeyVersionTemplate;
+    }
+
+    public void setCryptoKeyVersionTemplate(CryptoKeyVersionTemplate cryptoKeyVersionTemplate) {
+        this.cryptoKeyVersionTemplate = cryptoKeyVersionTemplate;
+    }
+
+    /**
+     * The labels of the crypto key.
+     */
+    @Updatable
+    public Map<String, String> getLabels() {
+        if (labels == null) {
+            labels = new HashMap<>();
+        }
+
+        return labels;
+    }
+
+    public void setLabels(Map<String, String> labels) {
+        this.labels = labels;
+    }
+
+    /**
+     * The ID of the primary key version.
+     */
+    @Updatable
+    public String getPrimaryKeyVersionId() {
+        return primaryKeyVersionId;
+    }
+
+    public void setPrimaryKeyVersionId(String primaryKeyVersionId) {
+        this.primaryKeyVersionId = primaryKeyVersionId;
+    }
+
+    /**
+     * The ID of the crypto key.
+     */
+    @Output
+    @Id
+    public String getId() {
+        return id;
+    }
+
+    public void setId(String id) {
+        this.id = id;
+    }
+
+    /**
+     * The list of versions of the crypto key.
+     */
+    @Output
+    public List<String> getVersions() {
+        if (versions == null) {
+            versions = new ArrayList<>();
+        }
+        return versions;
+    }
+
+    public void setVersions(List<String> versions) {
+        this.versions = versions;
+    }
+
+    @Override
+    public void copyFrom(CryptoKey model) throws Exception {
+        setId(model.getName());
+        setPurpose(model.getPurpose());
+        setKeyRing(findById(KeyRingResource.class, String.join("/", Arrays.copyOfRange(getId().split("/"), 0, 6))));
+        setName(getNameFromId());
+
+        if (model.hasNextRotationTime()) {
+            DateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy");
+            Date date = new Date(model.getNextRotationTime().getSeconds() * 1000L);
+            setNextRotationDate(dateFormat.format(date));
+        }
+
+        if (model.hasRotationPeriod()) {
+            setRotationPeriod(model.getRotationPeriod().getSeconds() / 86400);
+        }
+
+        if (model.getVersionTemplate() != null) {
+            CryptoKeyVersionTemplate cryptoKeyVersionTemplate = newSubresource(CryptoKeyVersionTemplate.class);
+            cryptoKeyVersionTemplate.copyFrom(model.getVersionTemplate());
+            setCryptoKeyVersionTemplate(cryptoKeyVersionTemplate);
+        }
+
+        if (model.hasPrimary()) {
+            setPrimaryKeyVersionId(model.getPrimary().getName().split("/")[9]);
+        }
+    }
+
+    @Override
+    protected boolean doRefresh() throws Exception {
+        KeyManagementServiceClient client = createClient(KeyManagementServiceClient.class);
+
+        CryptoKey cryptoKey = client.getCryptoKey(getId());
+
+        if (cryptoKey == null) {
+            return false;
+        }
+
+        copyFrom(cryptoKey);
+        refreshVersions(client);
```

Comment:
why is `refreshVersions` not part of copyFrom() ??

Reply:
From the method declaration, I thought `copyFrom` takes values from a given model, while `refreshVersions` needs access to the client to get another resource.

Reply:
copyFrom sets values based on the model, if that requires further api calls then the copyFrom does that as well, you just create a fresh client inside the copyFrom